lib/solvers/LongDistancePairSolver/LongDistancePairSolver.ts

```diff
@@ -0,0 +1,170 @@
+import { getConnectivityMapsFromInputProblem } from "lib/solvers/MspConnectionPairSolver/getConnectivityMapFromInputProblem"
+import type { MspConnectionPair } from "lib/solvers/MspConnectionPairSolver/MspConnectionPairSolver"
+import type {
+  InputProblem,
+  InputPin,
+  PinId,
+  InputChip,
+} from "lib/types/InputProblem"
+import { BaseSolver } from "../BaseSolver/BaseSolver"
+import { getAllPossiblePairs } from "lib/utils/get-all-possible-pairs"
+import { SchematicTraceSingleLineSolver2 } from "../SchematicTraceLinesSolver/SchematicTraceSingleLineSolver2/SchematicTraceSingleLineSolver2"
+import { doesTraceOverlapWithExistingTraces } from "lib/utils/does-trace-overlap-with-existing-traces"
+import { visualizeInputProblem } from "../SchematicTracePipelineSolver/visualizeInputProblem"
+import type { SolvedTracePath } from "../SchematicTraceLinesSolver/SchematicTraceLinesSolver"
+import type { ConnectivityMap } from "connectivity-map"
+
+export class LongDistancePairSolver extends BaseSolver {
+  public solvedLongDistanceTraces: SolvedTracePath[] = []
+  private queuedCandidatePairs: Array<
+    [InputPin & { chipId: string }, InputPin & { chipId: string }]
+  > = []
+  private currentCandidatePair:
+    | [InputPin & { chipId: string }, InputPin & { chipId: string }]
+    | null = null
+  private subSolver: SchematicTraceSingleLineSolver2 | null = null
+  private chipMap: Record<string, InputChip> = {}
+  private inputProblem: InputProblem
+  private netConnMap: ConnectivityMap
+
+  constructor(
+    private params: {
+      inputProblem: InputProblem
+      solvedTraces: SolvedTracePath[]
+      primaryMspConnectionPairs: MspConnectionPair[]
+    },
+  ) {
+    super()
+
+    const { inputProblem, primaryMspConnectionPairs } = this.params
+
+    this.inputProblem = inputProblem
+
+    // Pre-compute all possible long-distance pairs to check
+    const connectedPinIds = new Set<PinId>()
+    for (const pair of primaryMspConnectionPairs) {
+      connectedPinIds.add(pair.pins[0].pinId)
+      connectedPinIds.add(pair.pins[1].pinId)
+    }
+
+    const { netConnMap } = getConnectivityMapsFromInputProblem(inputProblem)
+    this.netConnMap = netConnMap
+    const pinMap = new Map<PinId, InputPin & { chipId: string }>()
+    for (const chip of inputProblem.chips) {
+      this.chipMap[chip.chipId] = chip
+      for (const pin of chip.pins) {
+        pinMap.set(pin.pinId, { ...pin, chipId: chip.chipId })
+      }
+    }
+
+    for (const netId of Object.keys(netConnMap.netMap)) {
+      const allPinIdsInNet = netConnMap.getIdsConnectedToNet(netId)
+      const unconnectedPinIds = allPinIdsInNet.filter(
+        (pinId) => !connectedPinIds.has(pinId),
+      )
+
+      if (unconnectedPinIds.length < 2) continue
+
+      const unconnectedPins = unconnectedPinIds
+        .map((pinId) => pinMap.get(pinId)!)
+        .filter(Boolean)
+
+      this.queuedCandidatePairs.push(...getAllPossiblePairs(unconnectedPins))
```

Comment:
hmm, this will have some issues. Each pin should only be able to connect to the N nearest pins. Otherwise this could become too complex since the number of pairs grows at N^2. So just look for the N=3 nearest pairs for each pin

Also it is a bit weird (maybe wrong?) that it's only considering unconnected pin -> unconnected pin. It could be valid to do unconnected pin -> connected pin